File: web/lib/workspace-access.ts
import { formatTimestamp } from "@/lib/runtime-presenters";

export const SESSION_COOKIE_NAME = "sevenflows_session";

export type WorkspaceMemberRole = "owner" | "admin" | "editor" | "viewer";

export type WorkspaceItem = {
  id: string;
  name: string;
  slug: string;
};

export type UserAccountItem = {
  id: string;
  email: string;
  display_name: string;
  status: string;
  last_login_at?: string | null;
};

export type WorkspaceMemberItem = {
  id: string;
  role: WorkspaceMemberRole;
  user: UserAccountItem;
  invited_by_user_id?: string | null;
  created_at: string;
  updated_at: string;
};

export type AuthSessionResponse = {
  token: string;
  workspace: WorkspaceItem;
  current_user: UserAccountItem;
  current_member: WorkspaceMemberItem;
  available_roles: WorkspaceMemberRole[];
  expires_at: string;
};

export type WorkspaceContextResponse = {
  workspace: WorkspaceItem;
  current_user: UserAccountItem;
  current_member: WorkspaceMemberItem;
  available_roles: WorkspaceMemberRole[];
  can_manage_members: boolean;
};

export function formatWorkspaceRole(role: WorkspaceMemberRole) {
  switch (role) {
    case "owner":
      return "所有者";
    case "admin":
      return "管理员";
    case "editor":
      return "编辑者";
    case "viewer":
      return "观察者";
    default:
      return role;
  }
}

export function canManageWorkspaceMembers(role: WorkspaceMemberRole) {
  return role === "owner" || role === "admin";
}

export function describeWorkspaceMemberActivity(member: WorkspaceMemberItem) {
  if (!member.user.last_login_at) {
    return "尚未登录";
  }

  return `最近登录 ${formatTimestamp(member.user.last_login_at)}`;
}

export function readSessionCookieFromDocument() {
  if (typeof document === "undefined") {
    return null;
  }

  const targetPrefix = `${SESSION_COOKIE_NAME}=`;
  const rawCookie = document.cookie
    .split(";")
    .map((item) => item.trim())
    .find((item) => item.startsWith(targetPrefix));

  if (!rawCookie) {
    return null;
  }

  return decodeURIComponent(rawCookie.slice(targetPrefix.length));
}

File: api/app/schemas/workspace_access.py
from datetime import datetime

from pydantic import BaseModel, Field

WorkspaceMemberRole = str


class WorkspaceItem(BaseModel):
    id: str
    name: str
    slug: str


class UserAccountItem(BaseModel):
    id: str
    email: str
    display_name: str
    status: str
    last_login_at: datetime | None = None


class WorkspaceMemberItem(BaseModel):
    id: str
    role: WorkspaceMemberRole
    user: UserAccountItem
    invited_by_user_id: str | None = None
    created_at: datetime
    updated_at: datetime


class AuthLoginRequest(BaseModel):
    email: str = Field(min_length=3, max_length=255)
    password: str = Field(min_length=6, max_length=128)


class AuthSessionResponse(BaseModel):
    token: str
    workspace: WorkspaceItem
    current_user: UserAccountItem
    current_member: WorkspaceMemberItem
    available_roles: list[WorkspaceMemberRole]
    expires_at: datetime


class WorkspaceContextResponse(BaseModel):
    workspace: WorkspaceItem
    current_user: UserAccountItem
    current_member: WorkspaceMemberItem
    available_roles: list[WorkspaceMemberRole]
    can_manage_members: bool


class WorkspaceMemberCreateRequest(BaseModel):
    email: str = Field(min_length=3, max_length=255)
    display_name: str = Field(min_length=1, max_length=128)
    password: str = Field(min_length=6, max_length=128)
    role: WorkspaceMemberRole = Field(default="viewer")


class WorkspaceMemberRoleUpdateRequest(BaseModel):
    role: WorkspaceMemberRole


class WorkspaceLogoutResponse(BaseModel):
    ok: bool = True
